### ISSUE TITLE
Dart compiler crashes when compiling tests if build directory is not empty

### ISSUE THREAD

**Issue by knaeckeKami** [flutter/flutter#48804]
I'm not sure if I should have posted this in the Repository for Dart instead, but since it only happens for me run using Flutter, I chose the Flutter repository.

If I run` flutter test`, then change the code of my application (under `lib/*` ), then run `flutter test` again, there is a ~50% chance that the tests fail with a crash of the dart compiler for my project.
See the output of `flutter test` in this case:

https://gist.github.com/knaeckeKami/2d51b57753da1729c18a9667145391a9

## How to Reproduce

Unfortunately, I was not able to reproduce a minimal sample where this issue occurs.
I hope the crashlog is helpful, though. 
Minimal flutter projects run fine, this issue only exists with my production-ready application.

## Workaround:

Running 
    
    rm -r ./build/*
fixes the problem temporarily. `flutter test` then runs successfully. But as soon as I change code under lib/ and run tests again, the issue will reappear. 

```
./flutterw doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.15.2 19C57, locale en-GB)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
[✓] Android Studio (version 3.5)
[!] IntelliJ IDEA Ultimate Edition (version 2019.1.2)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.41.0)
[!] Connected device
    ! No devices available

! Doctor found issues in 2 categories.
```

**Comment by mkustermann** [comment]
> https://gist.github.com/knaeckeKami/2d51b57753da1729c18a9667145391a9
```
00:02 +0: loading /Users/user/company/app-mobile/test/s3/s3_test.dart                                                                                                                                 Unhandled exception:
Crash when compiling file:///Users/user/company/app-mobile/test/medbase/medbase_risks_test.dart,
at character offset null:
NoSuchMethodError: The method '[]' was called on null.
Receiver: null
Tried calling: [](0)
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:53:5)
#1      computeVariance (package:front_end/src/fasta/kernel/type_algorithms.dart:95:42)
#2      computeVariance (package:front_end/src/fasta/kernel/type_algorithms.dart:149:17)
#3      SourceLibraryBuilder.computeVariances (package:front_end/src/fasta/source/source_library_builder.dart:2423:15)
#4      SourceLoader.computeVariances.<anonymous closure> (package:front_end/src/fasta/source/source_loader.dart:534:26)
#5      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:379:8)
#6      SourceLoader.computeVariances (package:front_end/src/fasta/source/source_loader.dart:532:14)
#7      KernelTarget.buildOutlines.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:265:14)
```

@jensjoha @johnniwinther Could you take a look?

**Comment by johnniwinther** [comment]
We're looking into it.

**Comment by jensjoha** [comment]
This should be fixed on the dart side with d7eeaed5dc3665b2c4e309141d885783c8e81e34. I'm not sure when it lands in flutter.
